r2-shared/Format/Format.ts

```diff
@@ -0,0 +1,147 @@
+import MediaType from "./MediaType";
+
+export interface IFormat {
+  name: string;
+  mediaType: MediaType;
+  fileExtension: string;
+}
```

Comment:
It’s more like I wanted to keep named parameters in static methods with more clarity as this can turn super ugly when done directly in `constructor`, cf. this [SO question](https://stackoverflow.com/questions/45507428/typescript-create-class-via-constructor-passing-in-named-parameters).

Reply:
@danielweck During the call there was a consensus in favor of using interfaces to create anonymous object arguments to replace the concept of "named parameters". But we're not sure about the naming of such interfaces. Is there a standard convention for this? Or a more idiomatic way to tackle the issue of named parameters?

Reply:
How about using a type alias on the object structure instead of an interface? This doesn't solve the naming issue, but at least it's clear that it's not a type to be implemented.

```typescript
type FormatDefinition = { name: string }

class Format {
  public name: string;

  constructor(params: FormatDefinition) {
    this.name = params.name;
  };

}
```

A few other ideas for the name: `FormatDefinition`, `FormatConstructor`, `FormatSchema`, `FormatData` and `FormatLike`, as seen in this PR.

Technically we could also ditch the name entirely. But I don't know enough TypeScript to see if that could be a problem:

```typescript
class Format {
  public name: string;

  constructor(params: { name: string }) {
    this.name = params.name;
  };

}
```

Reply:
I have come across a number of Javascript / TypeScript APIs where objects are used as function parameters in order to pass multiple atomic pieces of data (quick example from the top of my head: React component props and state), but there are many instances of the opposite design pattern (extreme example: curry-ing which “spreads” individual arguments over multiple function calls).

I think that generalising the use of objects to carry multiple function arguments is an unusual approach, but as far as I know it is not an anti-pattern ... though I am sure quite a few JS/TS programmers will find this “weird”, so I am not convinced this will improve the developer experience (DX) across the board.

Note that there is a runtime cost incurred when creating objects instead of just spreading the function arguments the “normal” way. Garbage collection in JS engines is pretty good, and the Readium SDK is not a “real time” / “high performance” system ... but still, this is something to keep in mind in addition to DX considerations.

Another potential problem is that variable with primitive types passed as function arguments are sometimes not used as constants within the closure scope, and instead are modified directly in order to avoid creating unnecessary stack allocations (i.e. this is an implementation detail below the API surface). With objects, there is the potential problem of mutating properties in one function scope, and inadvertently changing the value in another. Just a thought.

Reply:
Thanks for chiming in @danielweck. Yeah that would probably be overkill. If I understood correctly, @JayPanoz planned to use these mostly for large constructors with multiple parameters of the same type. One could argue that `Format` doesn't fit the bill, but it would probably be safer for `Link`, `Metadata`, etc. Because adding a new parameter constructors in an update means that the constructor could break silently if the type of the new parameter is the same.

Reply:
Yes, it is quite common in JavaScript/TypeScript APIs to use a "config" object argument which properties are assigned defaults when not defined on the caller's side. I can see how this would work for constructors or other initializers. That being said, it is also quite common to have such "config" function argument alongside other arguments, so the design pattern is quite open.

Personally I would favour consistency, so if "object with named properties is used to pass function arguments" gets applied to constructors, then I would prefer the design pattern to be used for all other public API functions (which then becomes quite cumbersome, as noted in a previous comment).